lib/aio/util.py

```diff
@@ -0,0 +1,142 @@
+# Copyright (C) 2024 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import argparse
+import asyncio
+import codecs
+import collections
+import contextlib
+import json
+import logging
+import ssl
+from typing import Any, AsyncIterator, Collection, Coroutine, Hashable, Mapping, Sequence, TypeVar
+
+import aiohttp
+
+from .jsonutil import JsonError, JsonObject, get_str, typechecked
+
+logger = logging.getLogger(__name__)
+
+# When mypy gets PEP 695 support: https://github.com/python/mypy/issues/15238
+# type JsonValue = None | bool | int | float | Sequence[JsonValue] | Mapping[str, JsonValue]
+K = TypeVar('K', bound=Hashable)
+T = TypeVar('T')
+V = TypeVar('V')
+
+
+class AsyncQueue(collections.deque[T]):
+    def __init__(self) -> None:
+        self._nonempty = asyncio.Event()
+        self._eof = False
+
+    async def next(self) -> T | None:
+        await self._nonempty.wait()
+        return self[0] if self else None
+
+    def done(self, item: T) -> None:
+        assert self.popleft() is item
+        if not self and not self._eof:
+            self._nonempty.clear()
+
+    def put(self, item: T) -> None:
+        self.append(item)
+        self._nonempty.set()
+
+    def eof(self) -> None:
+        self._nonempty.set()
+        self._eof = True
+
+
+# Simple LRU cache: when full, evict the least-recently `add()`-ed item.
+class LRUCache(dict[K, V]):
```

Comment:
How is that different from using https://docs.python.org/3/library/functools.html#functools.lru_cache directly in github.py?

Reply:
@functools.lru_cache applied to the `get()` method would have two problems:
 - @functools.lru_cache() on instance methods leaks (since the self parameter gets included as one of the "inputs").  Probably not relevant for us, since we have only one of them; but
 - this means that repeated `get()` calls on the same resource would always return the same value without bothering to check if it changed.  We actually *do* want to do the http request — we're polling after all — just with the etag/last-modified headers set.

Reply:
Sure, I wasn't proposing to apply this to the get() method, but to some `lru_cache(self.cache.__get__)` wrapping. But I did a quick google search and this doesn't actually seem to be that simple. Too bad, that feels like it'd be a nice stdlib addition.